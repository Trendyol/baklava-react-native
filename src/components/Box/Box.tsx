import { createBox } from '@ergenekonyigit/restyle';
import { Theme } from '../../theme';

const Box = createBox<Theme>();

export default Box;
