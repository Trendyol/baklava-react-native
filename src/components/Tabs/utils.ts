export default function insertObjectBetweenElements<T>(arr: T[], obj: T): T[] {
  if (arr.length <= 1) {
    return arr;
  }
  const result: T[] = [];

  for (let i = 0; i < arr.length - 1; i++) {
    result.push(arr[i], obj);
  }
  result.push(arr[arr.length - 1]);

  return result;
}
