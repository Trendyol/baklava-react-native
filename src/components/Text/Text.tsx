import { createText } from '@shopify/restyle';
import { Theme } from '../../theme';

const Text = createText<Theme>();

export default Text;
