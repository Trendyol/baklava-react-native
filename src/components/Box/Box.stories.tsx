import { ComponentStory } from '@storybook/react-native';
import React from 'react';
import Text from '../Text/Text';
import Box from './Box';

export default {
  title: 'Box',
  component: Box,
};

type BoxStory = ComponentStory<typeof Box>;

export const Basic: BoxStory = args => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Box
    </Text>
    <Box {...args} />
  </>
);

Basic.args = {
  backgroundColor: 'primaryKey',
  width: 200,
  height: 200,
};
