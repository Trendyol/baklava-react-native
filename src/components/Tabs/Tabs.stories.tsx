import { ComponentMeta, ComponentStory } from '@storybook/react-native';
import React, { useState } from 'react';
import Box from '../Box/Box';
import Text from '../Text/Text';
import Tabs from './Tabs';
import { useWindowDimensions } from 'react-native';

const TabsMeta: ComponentMeta<typeof Tabs> = {
  title: 'Tabs',
  component: Tabs,
  argTypes: {
    onValueChange: { action: 'pressed the button' },
  },
  args: {
    value: 'tab1',
    defaultValue: 'tab1',
  },
};

export default TabsMeta;

type TabsStory = ComponentStory<typeof Tabs>;

export const Basic: TabsStory = args => {
  const [value, setValue] = React.useState('');
  React.useEffect(() => {
    setValue(args.value);
  }, [args.value]);
  return (
    <>
      <Text p="2xs" variant="subtitle01Bold">
        Tabs
      </Text>
      <Tabs
        {...args}
        onValueChange={setValue}
        value={value}
        defaultValue="tab1">
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab Title" caption="Caption" />
          <Tabs.Option value="tab2" title="Tab Title" />
        </Tabs.List>
        <Tabs.Content value="tab1">
          <Box mt="s">
            <Text>Tab 1 Content</Text>
          </Box>
        </Tabs.Content>
        <Tabs.Content value="tab2">
          <Box mt="s">
            <Text>Tab 2 Content</Text>
          </Box>
        </Tabs.Content>
      </Tabs>
    </>
  );
};

export const TabTypes: TabsStory = () => {
  return (
    <>
      <Text p="2xs" variant="subtitle01Bold">
        Tab Types
      </Text>
      <Box px="m" py="2xs">
        <Tabs value={'tab1'} onValueChange={() => {}} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab1" title="Active Tab" caption="Caption" />
            <Tabs.Option value="tab2" title="Inactive Tab" caption="Caption" />
            <Tabs.Option
              value="tab2"
              title="Disabled Tab"
              caption="Caption"
              disabled
            />
          </Tabs.List>
        </Tabs>
      </Box>
    </>
  );
};

export const Cases: TabsStory = () => {
  const [value, setValue] = useState('tab1');
  const [value2, setValue2] = useState('tab1');
  const [value3, setValue3] = useState('tab1');
  const [value4, setValue4] = useState('tab1');

  return (
    <>
      <Text p="2xs" variant="subtitle01Bold">
        Tab Cases
      </Text>
      <Box px="m" py="2xs">
        <Tabs value={value} onValueChange={setValue} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab1" title="Tab Title" caption="Caption" />
            <Tabs.Option value="tab2" title="Tab Title" />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>

      <Box px="m" py="2xs" mt="5xl">
        <Tabs value={value2} onValueChange={setValue2} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab1" title="Tab Title" caption="Caption" />
            <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
            <Tabs.Option
              value="tab3"
              title="Tab Title"
              caption="Caption"
              disabled
            />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab3">
            <Box mt="s">
              <Text>Tab 3 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>

      <Box px="m" py="2xs" mt="5xl">
        <Tabs value={value3} onValueChange={setValue3} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option
              value="tab1"
              title="Lorem ipsum dolor sit amet"
              caption="Caption"
            />
            <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>

      <Box px="m" py="2xs" mt="5xl">
        <Tabs value={value4} onValueChange={setValue4} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab1" title="Tab Title" caption="Caption" />
            <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
            <Tabs.Option value="tab3" title="Tab Title" />
            <Tabs.Option value="tab4" title="Tab Title" caption="Caption" />
            <Tabs.Option value="tab5" title="Tab Title" caption="Caption" />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab3">
            <Box mt="s">
              <Text>Tab 3 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab4">
            <Box mt="s">
              <Text>Tab 4 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab5">
            <Box mt="s">
              <Text>Tab 5 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>
    </>
  );
};

export const IconAndBadge: TabsStory = () => {
  const [value1, setValue1] = useState('tab1');
  const [value2, setValue2] = useState('tab1');
  const [value3, setValue3] = useState('tab1');

  return (
    <>
      <Text p="2xs" variant="subtitle01Bold">
        Icon And Badge Tab Cases
      </Text>

      <Box px="m" py="2xs" mt="5xl">
        <Tabs value={value1} onValueChange={setValue1} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
            <Tabs.Option
              value="tab1"
              title="Tab Title"
              caption="Caption"
              iconName="heart"
            />
            <Tabs.Option value="tab3" title="Tab Title" />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab3">
            <Box mt="s">
              <Text>Tab 3 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab4">
            <Box mt="s">
              <Text>Tab 4 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab5">
            <Box mt="s">
              <Text>Tab 5 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>

      <Box px="m" py="2xs" mt="5xl">
        <Tabs value={value2} onValueChange={setValue2} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
            <Tabs.Option
              value="tab1"
              title="Tab Title"
              caption="Caption"
              badgeText="New"
            />
            <Tabs.Option value="tab3" title="Tab Title" />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab3">
            <Box mt="s">
              <Text>Tab 3 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab4">
            <Box mt="s">
              <Text>Tab 4 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab5">
            <Box mt="s">
              <Text>Tab 5 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>

      <Box px="m" py="2xs" mt="5xl">
        <Tabs value={value3} onValueChange={setValue3} defaultValue="tab1">
          <Tabs.List>
            <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
            <Tabs.Option
              value="tab1"
              title="Tab Title"
              caption="Caption"
              iconName="confetti"
              badgeText="Ipsum"
            />
            <Tabs.Option value="tab3" title="Tab Title" />
          </Tabs.List>
          <Tabs.Content value="tab1">
            <Box mt="s">
              <Text>Tab 1 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab2">
            <Box mt="s">
              <Text>Tab 2 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab3">
            <Box mt="s">
              <Text>Tab 3 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab4">
            <Box mt="s">
              <Text>Tab 4 Content</Text>
            </Box>
          </Tabs.Content>
          <Tabs.Content value="tab5">
            <Box mt="s">
              <Text>Tab 5 Content</Text>
            </Box>
          </Tabs.Content>
        </Tabs>
      </Box>
    </>
  );
};

export const Example: TabsStory = () => {
  const [value, setValue] = useState('tab1');
  const dims = useWindowDimensions();

  return (
    <Box height={dims.height} backgroundColor="dangerContrast">
      <Tabs value={value} onValueChange={setValue} defaultValue="tab1">
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab Title" caption="Caption" />
          <Tabs.Option value="tab2" title="Tab Title" caption="Caption" />
        </Tabs.List>
        <Tabs.Content flex={1} value="tab1" backgroundColor="infoKey">
          <Box mt="s">
            <Text>Tab 1 Content</Text>
          </Box>
        </Tabs.Content>
        <Tabs.Content flex={1} value="tab2" backgroundColor="infoKey">
          <Box mt="s">
            <Text>Tab 2 Content</Text>
          </Box>
        </Tabs.Content>
      </Tabs>
    </Box>
  );
};
