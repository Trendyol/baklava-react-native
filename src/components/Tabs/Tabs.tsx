import React, {
  createContext,
  PropsWithChildren,
  useContext,
  useEffect,
  useState,
} from 'react';
import { FlatList, TouchableOpacity } from 'react-native';
import Box, { BoxProps } from '../Box/Box';
import Text from '../Text/Text';
import Icon from '../Icon/Icon';
import insertObjectBetweenElements from './utils';
import Badge from '../Badge/Badge';
import { IconNameType } from '../Icon/types';

type TabsContextType = {
  value: string;
  onValueChange: (newValue: string) => void;
};

const TabsContext = createContext<TabsContextType | null>(null);

type TabsProps = BoxProps & {
  value: string;
  defaultValue?: string;
  onValueChange: (value: string) => void;
};

const Tabs = ({
  value: controlledValue,
  defaultValue,
  onValueChange,
  ...rest
}: TabsProps) => {
  const [value, setValue] = useState(defaultValue ?? '');

  useEffect(() => {
    if (controlledValue !== undefined) {
      setValue(controlledValue);
    }
  }, [controlledValue]);

  const handleChange = (newValue: string) => {
    if (controlledValue === undefined) {
      setValue(newValue);
    }
    onValueChange(newValue);
  };

  return (
    <TabsContext.Provider value={{ value, onValueChange: handleChange }}>
      <Box testID="tabsBox" accessibilityLabel="tabsBox" flex={1} {...rest} />
    </TabsContext.Provider>
  );
};

const List = ({ children }: PropsWithChildren) => {
  const data = React.useMemo(() => {
    const childrenArray = React.Children.toArray(children);
    return insertObjectBetweenElements(
      childrenArray,
      <Box
        flex={1}
        mx="m"
        width={1}
        height={30}
        alignSelf="center"
        backgroundColor="neutralLighter"
      />,
    );
  }, [children]);

  return (
    <Box testID="tabsListBox" accessibilityLabel="tabsListBox">
      <FlatList
        data={data}
        testID="tabsFlatList"
        accessibilityLabel="tabsFlatList"
        keyExtractor={(item, index) => index.toString()}
        horizontal
        showsHorizontalScrollIndicator={false}
        renderItem={({ item }) => item as React.JSX.Element}
        contentContainerStyle={{ flexGrow: 1, justifyContent: 'space-around' }}
      />
    </Box>
  );
};

type OptionProps = {
  value: string;
  title: string;
  caption?: string;
  disabled?: boolean;
  iconName?: IconNameType;
  badgeText?: string;
};

const Option = ({
  value,
  title,
  caption = '',
  disabled = false,
  iconName = undefined,
  badgeText = undefined,
}: OptionProps) => {
  const context = useContext(TabsContext);
  if (!context) {
    throw new Error('Tabs.Option must be used within Tabs');
  }

  const isSelected = context.value === value;
  const conditionalColor = disabled
    ? 'neutralLight'
    : isSelected
    ? 'primaryKey'
    : 'neutralDarker';

  const underline = React.useMemo(() => {
    if (!isSelected) {
      return null;
    }

    return (
      <Box
        position="absolute"
        bottom={0}
        alignSelf="center"
        width="100%"
        borderBottomWidth={2}
        borderColor="primaryKey"
      />
    );
  }, [isSelected]);

  return (
    <TouchableOpacity
      disabled={disabled}
      style={{ justifyContent: 'center' }}
      onPress={() => !disabled && context.onValueChange(value)}>
      <Box
        testID="tabOptionBox"
        alignItems="center"
        justifyContent="center"
        paddingVertical="xs">
        <Box flexDirection="row" alignItems="center">
          {iconName ? (
            <Box pr="3xs" testID="tabIconBox">
              <Icon
                name={iconName}
                size="s"
                color={conditionalColor}
                testID="tabIcon"
              />
            </Box>
          ) : null}
          <Box testID="tabTitleBox">
            <Text
              testID="tabTitle"
              fontWeight="500"
              color={conditionalColor}
              numberOfLines={1}
              ellipsizeMode="tail">
              {title}
            </Text>
          </Box>
          {badgeText ? (
            <Box pl="3xs" testID="tabBadgeBox">
              <Badge
                testID="tabBadge"
                text={badgeText}
                size="small"
                transparent
                color="neutralFull"
                backgroundColor="dangerKey"
              />
            </Box>
          ) : null}
        </Box>
        {caption ? (
          <Text
            testID="tabCaption"
            fontSize={12}
            color={disabled ? 'neutralLight' : 'neutralDarker'}>
            {caption}
          </Text>
        ) : null}
      </Box>
      {underline}
    </TouchableOpacity>
  );
};

type ContentProps = BoxProps & {
  value: string;
};

const Content = ({ value, ...rest }: ContentProps) => {
  const context = useContext(TabsContext);
  if (!context) {
    throw new Error('Tabs.Content must be used within Tabs');
  }

  return context.value === value ? (
    <Box
      testID="tabsContentBox"
      accessibilityLabel="tabsContentBox"
      {...rest}
    />
  ) : null;
};

Tabs.List = List;
Tabs.Option = Option;
Tabs.Content = Content;

export default Tabs;
