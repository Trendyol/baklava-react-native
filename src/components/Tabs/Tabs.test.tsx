import React from 'react';
import theme from '../../theme';
import { fireEvent, render } from '../../test-utils';
import Tabs from './Tabs';
import Box from '../Box/Box';
import Text from '../Text/Text';
import insertObjectBetweenElements from './utils';

describe('Tabs', () => {
  beforeEach(() => {});

  test('should render Tabs correctly', () => {
    // given
    const { toJSON, getByTestId } = render(
      <Tabs value="tab1" onValueChange={() => {}} defaultValue="tab1">
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
          <Tabs.Option value="tab2" title="Tab 2" />
        </Tabs.List>
        <Tabs.Content value="tab1">
          <Text>Tab 1 Content</Text>
        </Tabs.Content>
        <Tabs.Content value="tab2">
          <Text>Tab 2 Content</Text>
        </Tabs.Content>
      </Tabs>,
    );

    // when
    const tabsComponent = getByTestId('tabsBox');
    const listComponent = getByTestId('tabsListBox');

    // then
    expect(toJSON()).toMatchSnapshot();
    expect(tabsComponent).toBeTruthy();
    expect(listComponent).toBeTruthy();
  });

  test('should change selected tab on press', () => {
    // given
    const onValueChange = jest.fn();
    const { getByText } = render(
      <Tabs value="tab1" onValueChange={onValueChange} defaultValue="tab1">
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
          <Tabs.Option value="tab2" title="Tab 2" />
        </Tabs.List>
        <Tabs.Content value="tab1">
          <Text>Tab 1 Content</Text>
        </Tabs.Content>
        <Tabs.Content value="tab2">
          <Text>Tab 2 Content</Text>
        </Tabs.Content>
      </Tabs>,
    );

    // when
    fireEvent.press(getByText('Tab 2'));

    // then
    expect(onValueChange).toHaveBeenCalledWith('tab2');
  });

  test('should change selected tab on press when value undefined', () => {
    // given
    const onValueChange = jest.fn();
    const { getByText } = render(
      <Tabs value={undefined} onValueChange={onValueChange} defaultValue="tab1">
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
          <Tabs.Option value="tab2" title="Tab 2" />
        </Tabs.List>
        <Tabs.Content value="tab1">
          <Text>Tab 1 Content</Text>
        </Tabs.Content>
        <Tabs.Content value="tab2">
          <Text>Tab 2 Content</Text>
        </Tabs.Content>
      </Tabs>,
    );

    // when
    fireEvent.press(getByText('Tab 2'));

    // then
    expect(onValueChange).toHaveBeenCalledWith('tab2');
  });

  test('should not trigger onValueChange when disabled tab is pressed', () => {
    // given
    const onValueChange = jest.fn();
    const { getByText } = render(
      <Tabs value="tab1" onValueChange={onValueChange}>
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
          <Tabs.Option value="tab2" title="Tab 2" disabled />
        </Tabs.List>
        <Tabs.Content value="tab1">
          <Text>Tab 1 Content</Text>
        </Tabs.Content>
        <Tabs.Content value="tab2">
          <Text>Tab 2 Content</Text>
        </Tabs.Content>
      </Tabs>,
    );

    // when
    fireEvent.press(getByText('Tab 2'));

    // then
    expect(onValueChange).not.toHaveBeenCalled();
  });

  test('should update value state when uncontrolled (defaultValue is used)', () => {
    // given
    const onValueChangeMock = jest.fn();
    const { getByText } = render(
      <Tabs value="tab1" defaultValue="tab1" onValueChange={onValueChangeMock}>
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
          <Tabs.Option value="tab2" title="Tab 2" />
        </Tabs.List>
      </Tabs>,
    );

    const tab2 = getByText('Tab 2');

    // when
    fireEvent.press(tab2);

    // then
    expect(onValueChangeMock).toHaveBeenCalledWith('tab2');
  });

  test('should render correct content based on selected tab', () => {
    // given
    const { getByTestId, queryByTestId } = render(
      <Tabs value="tab1" onValueChange={() => {}}>
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
          <Tabs.Option value="tab2" title="Tab 2" />
        </Tabs.List>
        <Tabs.Content value="tab1">
          <Box testID="content-tab1">
            <Text>Tab 1 Content</Text>
          </Box>
        </Tabs.Content>
        <Tabs.Content value="tab2">
          <Box testID="content-tab2">
            <Text>Tab 2 Content</Text>
          </Box>
        </Tabs.Content>
      </Tabs>,
    );

    // then
    expect(getByTestId('content-tab1')).toBeTruthy();
    expect(queryByTestId('content-tab2')).toBeNull();
  });

  test('should have neutralLight color when disabled', () => {
    // given
    const { getByTestId } = render(
      <Tabs value="tab1" onValueChange={jest.fn()}>
        <Tabs.List>
          <Tabs.Option
            value="tab1"
            title="Tab 1"
            disabled
            iconName="heart"
            badgeText="test"
          />
        </Tabs.List>
      </Tabs>,
    );

    // when
    const title = getByTestId('tabTitle');
    const icon = getByTestId('tabIcon');
    const badge = getByTestId('tabBadge');
    const badgeText = getByTestId('badgeText');

    // then
    expect(title.props.style[0].color).toEqual(theme.colors.neutralLight);
    expect(icon.props.fill).toEqual(theme.colors.neutralLight);
    expect(badgeText.props.style[0].color).toEqual(theme.colors.neutralFull);
    expect(badge.props.style[1].backgroundColor).toEqual(
      theme.colors.dangerKey,
    );
  });

  test('should have primaryKey color when selected and enabled', () => {
    // given
    const { getByTestId } = render(
      <Tabs value="tab1" onValueChange={jest.fn()}>
        <Tabs.List>
          <Tabs.Option
            value="tab1"
            title="Tab 1"
            iconName="heart"
            badgeText="test"
          />
        </Tabs.List>
      </Tabs>,
    );

    // when
    const title = getByTestId('tabTitle');
    const icon = getByTestId('tabIcon');
    const badge = getByTestId('tabBadge');
    const badgeText = getByTestId('badgeText');

    // then
    expect(title.props.style[0].color).toEqual(theme.colors.primaryKey);
    expect(icon.props.fill).toEqual(theme.colors.primaryKey);
    expect(badgeText.props.style[0].color).toEqual(theme.colors.neutralFull);
    expect(badge.props.style[1].backgroundColor).toEqual(
      theme.colors.dangerKey,
    );
  });

  test('should render caption when provided', () => {
    // given
    const { getByText } = render(
      <Tabs value="tab1" onValueChange={jest.fn()}>
        <Tabs.List>
          <Tabs.Option
            value="tab1"
            title="Tab 1"
            caption="Test Caption"
            disabled
          />
        </Tabs.List>
      </Tabs>,
    );

    // when
    const tab1 = getByText('Tab 1');

    // then
    expect(getByText('Test Caption')).toBeTruthy();
    expect(tab1.props.style[0].color).toEqual(theme.colors.neutralLight);
  });

  test('should not render caption when not provided', () => {
    // given
    const { queryByText } = render(
      <Tabs value="tab1" onValueChange={jest.fn()}>
        <Tabs.List>
          <Tabs.Option value="tab1" title="Tab 1" />
        </Tabs.List>
      </Tabs>,
    );

    // then
    expect(queryByText('Test Caption')).toBeNull();
  });

  test('should render title with neutralDarker color when not disabled and not selected', () => {
    // Given
    const onValueChange = jest.fn();
    const { getByText } = render(
      <Tabs value="tab1" onValueChange={onValueChange}>
        <Tabs.List>
          <Tabs.Option value="tab1" title="Option 1" caption="Caption 1" />
          <Tabs.Option value="tab2" title="Option 2" caption="Caption 2" />
        </Tabs.List>
      </Tabs>,
    );

    // When
    const option2Text = getByText('Option 2');

    // Then
    expect(option2Text.props.style[0].color).toBe(theme.colors.neutralDarker);
  });

  test('should throw error if Tabs.Option is used outside of Tabs', () => {
    // given
    const consoleErrorSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    // then
    expect(() => render(<Tabs.Option value="tab1" title="Tab 1" />)).toThrow(
      'Tabs.Option must be used within Tabs',
    );

    consoleErrorSpy.mockRestore();
  });

  test('should throw error if Tabs.Content is used outside of Tabs', () => {
    // given
    const consoleErrorSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    // then
    expect(() => render(<Tabs.Content value="tab1" />)).toThrow(
      'Tabs.Content must be used within Tabs',
    );

    consoleErrorSpy.mockRestore();
  });

  test('provider', () => {});
});

describe('insertObjectBetweenElements', () => {
  test('should return the same array if it has 0 elements', () => {
    // given
    const input: string[] = [];
    const separator = '|';

    const result = insertObjectBetweenElements(input, separator);

    // then
    expect(result).toEqual([]);
  });

  test('should return the same array if it has 1 element', () => {
    // given
    const input = ['A'];
    const separator = '|';

    const result = insertObjectBetweenElements(input, separator);

    // then
    expect(result).toEqual(['A']);
  });

  test('should insert the object between elements', () => {
    // given
    const input = ['A', 'B', 'C'];
    const separator = '|';

    const result = insertObjectBetweenElements(input, separator);

    // then
    expect(result).toEqual(['A', '|', 'B', '|', 'C']);
  });
});
