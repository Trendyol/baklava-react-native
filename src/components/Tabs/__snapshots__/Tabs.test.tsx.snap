// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render Tabs correctly 1`] = `
<View
  accessibilityLabel="tabsBox"
  accessible={true}
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
  testID="tabsBox"
>
  <View
    accessibilityLabel="tabsListBox"
    accessible={true}
    style={
      [
        {},
      ]
    }
    testID="tabsListBox"
  >
    <RCTScrollView
      accessibilityLabel="tabsFlatList"
      contentContainerStyle={
        {
          "flexGrow": 1,
          "justifyContent": "space-around",
        }
      }
      data={
        [
          <Option
            title="Tab 1"
            value="tab1"
          />,
          <ForwardRef
            alignSelf="center"
            backgroundColor="neutralLighter"
            flex={1}
            height={30}
            mx="m"
            width={1}
          />,
          <Option
            title="Tab 2"
            value="tab2"
          />,
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      showsHorizontalScrollIndicator={false}
      stickyHeaderIndices={[]}
      testID="tabsFlatList"
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <View
              accessibilityLabel="tabOptionBox"
              accessible={true}
              style={
                [
                  {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "paddingVertical": 12,
                  },
                ]
              }
              testID="tabOptionBox"
            >
              <View
                accessible={false}
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  accessibilityLabel="tabTitleBox"
                  accessible={true}
                  style={
                    [
                      {},
                    ]
                  }
                  testID="tabTitleBox"
                >
                  <Text
                    accessibilityLabel="tabTitle"
                    accessible={true}
                    ellipsizeMode="tail"
                    numberOfLines={1}
                    style={
                      [
                        {
                          "color": "#F27A1A",
                          "fontFamily": "Rubik-Regular",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "textAlign": "left",
                        },
                      ]
                    }
                    testID="tabTitle"
                  >
                    Tab 1
                  </Text>
                </View>
              </View>
            </View>
            <View
              accessible={false}
              style={
                [
                  {
                    "alignSelf": "center",
                    "borderBottomWidth": 2,
                    "borderColor": "#F27A1A",
                    "bottom": 0,
                    "position": "absolute",
                    "width": "100%",
                  },
                ]
              }
            />
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessible={false}
            style={
              [
                {
                  "alignSelf": "center",
                  "backgroundColor": "#AFBBCA",
                  "flex": 1,
                  "height": 30,
                  "marginHorizontal": 16,
                  "width": 1,
                },
              ]
            }
          />
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <View
              accessibilityLabel="tabOptionBox"
              accessible={true}
              style={
                [
                  {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "paddingVertical": 12,
                  },
                ]
              }
              testID="tabOptionBox"
            >
              <View
                accessible={false}
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  accessibilityLabel="tabTitleBox"
                  accessible={true}
                  style={
                    [
                      {},
                    ]
                  }
                  testID="tabTitleBox"
                >
                  <Text
                    accessibilityLabel="tabTitle"
                    accessible={true}
                    ellipsizeMode="tail"
                    numberOfLines={1}
                    style={
                      [
                        {
                          "color": "#273142",
                          "fontFamily": "Rubik-Regular",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "textAlign": "left",
                        },
                      ]
                    }
                    testID="tabTitle"
                  >
                    Tab 2
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessibilityLabel="tabsContentBox"
    accessible={true}
    style={
      [
        {},
      ]
    }
    testID="tabsContentBox"
  >
    <Text
      accessible={false}
      style={
        [
          {
            "color": "#273142",
            "fontFamily": "Rubik-Regular",
            "fontSize": 14,
            "textAlign": "left",
          },
        ]
      }
    >
      Tab 1 Content
    </Text>
  </View>
</View>
`;
